except_star.md

```diff
@@ -483,6 +483,99 @@ errors is error prone.
 
 We can consider allowing some of them in future versions of Python.
 
+## Traceback Groups
+
+For regular exceptions, the traceback represents a simple path of frames, from
+the frame in which the exception was raised to the frame in which it was
+was caught or, if it hasn't been caught yet, the frame that the program's
+execution is currently in.
+At any point in time the interpreter has at most one 'current exception' (the
+one returned by `sys.exc_info()`), and as execution progresses, it is
+necessary to append frames to this exception's traceback. Therefore the
+traceback's list of frames is stored in reverse order, such that the `next`
+links lead from the last frame to the first. Appending a new frame is then
+simply a matter of inserting a new head to the linked list referenced from
+the exception's `__traceback__` field.
+
+When we add `ExceptionGroup`s, the traceback becomes a tree of frames: each
+exception has a traceback that describes the path from where it was raised
+to where it was first added to an `ExceptionGroup`. From that point, all
+exceptions in the same group travel through the same frames, possibly
+merging with other exceptions into larger `ExceptionGroup`s. In this section
+we describe the `TracebackGroup` data structure which we are proposing to
+represent the tracebacks of the exceptions in an `ExceptionGroup`.
+
+`TracebackGroup`s satisfy the following requirements:
+1. Supports appending a frame in O(1) time.
+2. Supports efficient insertions and deletions of tracebacks as the
+exception group grows and splits.
+3. Does not hold references that can prevent GC of exceptions from the group.
+4. Does not require API breaking changes to the current traceback
+representation for regular exceptions.
+
+The `TracebackGroup` type is a subclass of `Traceback` and therefore can be
+assigned to the `__traceback__` field of an exception.
+Instead of a single `tb_next` link to a single traceback, the
+`TracebackGroup` has a `tb_next_map` that maps each exception from the
+`ExceptionGroup` to a traceback. This map holds weak references to
+its keys so that it does not prevent GC of exceptions.
+
+
+**Appending a frame** `f` to the end of the traceback of an exception `eg`
+is exactly the same as for traceback groups and regular tracebacks:
+
+```python
+tb = Traceback(tb_frame = f, tb_next = eg.__traceback__)
+eg.__traceback__ = tb
+```
+
+**Adding an exception** `e` to the `ExceptionGroup eg` requires adding its
+exception(s) to `tb_next_map` of `eg`:
+
+```python
+if not isinstance(e, ExceptionGroup):
+    keys = [e]                               # e is a simple exception
+else:
+    keys = list(e.tb_next_map.keys())        # e is an ExceptionGroup
+
+for k in keys:
+    if not isinstance(eg.__traceback__, TracebackGroup):
+        # simple traceback, convert it to a group    (WILL WE EVER NEED THIS?)
+        eg.__traceback__ = TracebackGroup.fromTraceback(eg.__traceback__)
```

Comment:
I'm not at all sure that this 'TracebackGroup.fromTraceback' call can ever happen - I put it in for completion, so that we can hopefully rule it out.

Why would we add an exception to an ExceptionGroup that was created in another frame? Wouldn't that mean that we create a new ExceptionGroup in the current frame and insert our exception along with the original exception group, so that we end up with the original EG nested in the new one?

Reply:
Maybe we should first write down the spec for ExceptionGroup and how it is manipulated during a try-except* block. Then we will know what invariants we can assume for the TracebackGroup operations.

Reply:
> Maybe we should first write down the spec for ExceptionGroup and how it is manipulated during a try-except* block. 

+1. It's the ExceptionGroup API and how they're used by try..except* that will guide the TracebackGroup design, not the other way around.